src/Microsoft.Dnx.Tooling/Utils/PackageFeedCache.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Dnx.Tooling.Restore.NuGet;
+using NuGet;
+using System.Collections.Generic;
+
+namespace Microsoft.Dnx.Tooling
+{
+    public class PackageFeedCache
+    {
+        private readonly ConcurrentDictionary<PackageSource, IPackageFeed> _packageFeeds = new ConcurrentDictionary<PackageSource, IPackageFeed>(new PackageSourceComparer());
+
+        public IPackageFeed GetPackageFeed(PackageSource source, bool noCache, bool ignoreFailedSources,
+            Reports reports)
+        {
+            return _packageFeeds.GetOrAdd(source, CreatePackageFeed(source, noCache, ignoreFailedSources, reports));
+        }
+
+        private IPackageFeed CreatePackageFeed(PackageSource source, bool noCache, bool ignoreFailedSources,
+            Reports reports)
+        {
+            // Check if the feed is a file path
+            if (source.IsLocalFileSystem())
+            {
+                return PackageFolderFactory.CreatePackageFolderFromPath(source.Source, ignoreFailedSources, reports);
+            }
+            else
+            {
+                var httpSource =  new HttpSource(
+                    source.Source,
+                    source.UserName,
+                    source.Password,
+                    reports);
+
+                Uri packageBaseAddress;
+                if (NuGetv3Feed.DetectNuGetV3(httpSource, noCache, out packageBaseAddress))
+                {
+                    if (packageBaseAddress == null)
+                    {
+                        reports.Information.WriteLine(
+                            $"Ignoring NuGet v3 feed {source.Source.Yellow().Bold()}, which doesn't provide PackageBaseAddress resource.");
+                        return null;
+                    }
+
+                    httpSource = new HttpSource(
+                        packageBaseAddress.AbsoluteUri,
+                        source.UserName,
+                        source.Password,
+                        reports);
+
+                    return new NuGetv3Feed(
+                        httpSource,
+                        noCache,
+                        reports,
+                        ignoreFailedSources);
+                }
+
+                return new NuGetv2Feed(
+                    httpSource,
+                    noCache,
+                    reports,
+                    ignoreFailedSources);
+            }
+        }
+    }
+
+    internal class PackageSourceComparer : IEqualityComparer<PackageSource>
+    {
+        public bool Equals(PackageSource left, PackageSource right)
+        {
+            return string.Equals(left.Source, right.Source) &&
```

Comment:
Spoke with @BrennanConroy offline - `StringComparison.Ordinal` might be more appropriate here.